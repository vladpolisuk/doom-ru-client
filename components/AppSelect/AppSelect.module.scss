.app_select {
	display: flex;
	flex-direction: column;
	align-items: flex-start;
	justify-content: center;
	width: 100%;
	border: 0;
	font-size: 1rem;
	color: var(--color);
	background-color: var(--color-6);
	padding: 8px 8px;
	border-radius: 0.25rem;

	&_option {
		color: var(--color);
		background-color: var(--background);
	}
}

.app_select--invalid {
	box-shadow: 0 0 0 2px rgb(255, 29, 29);

	&:focus-visible {
		box-shadow: 0 0 0 2px rgb(255, 29, 29);
	}
}
