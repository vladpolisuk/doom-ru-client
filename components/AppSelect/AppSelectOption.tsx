import { FC, memo, OptionHTMLAttributes } from 'react';
import s from './AppSelect.module.scss';

export type IAppSelectOption = FC<OptionHTMLAttributes<HTMLOptionElement>>;

const AppSelectOption: IAppSelectOption = ({ children, ...props }) => {
	return (
		<option
			data-testid='app-select-option'
			className={s.app_select_option}
			{...props}>
			{children}
		</option>
	);
};

export default AppSelectOption;
