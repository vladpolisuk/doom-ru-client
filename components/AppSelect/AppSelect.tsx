import { FC, forwardRef, memo, OptionHTMLAttributes } from 'react';
import { BaseAppComponent } from '../../types/components';
import getConcatenatedStylesByCondition from '../../utils/ui/getConcatenatedStylesByCondition';
import s from './AppSelect.module.scss';
import AppSelectOption from './AppSelectOption';

type IAppSelect = BaseAppComponent<HTMLSelectElement>;

type IAppSelectCompound = {
	Option: FC<OptionHTMLAttributes<HTMLOptionElement>>;
};

/**
 * The common select component in the application
 * @type `FC<AppSelect>`
 * @memo `true`
 * @return `html:select`
 */
// @ts-ignore
const AppSelect: FC<IAppSelect> & IAppSelectCompound = memo(
	forwardRef<HTMLSelectElement, IAppSelect>((props, ref) => {
		const { children, className = '', resetStyles = false, defaultValue, ...extra } = props;
		const selectStyles = getConcatenatedStylesByCondition(resetStyles, className, s.app_select);

		return (
			<select
				ref={ref}
				data-testid='app-select'
				className={selectStyles}
				defaultValue={defaultValue}
				{...extra}>
				{children}
			</select>
		);
	})
);

AppSelect.displayName = 'AppSelect';
AppSelect.Option = memo(AppSelectOption);
export default AppSelect;
