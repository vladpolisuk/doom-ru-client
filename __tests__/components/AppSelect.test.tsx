import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import AppSelect from '../../components/AppSelect/AppSelect';

describe('AppSelect', () => {
	const data = [
		{ value: '', label: 'Choose option' },
		{ value: '1', label: 'Option 1' },
		{ value: '2', label: 'Option 2' },
		{ value: '3', label: 'Option 3' }
	];

	it('should render AppSelect', () => {
		const wrapper = render(
			<AppSelect>
				{data.map(option => (
					<AppSelect.Option
						key={option.value}
						value={option.value}>
						{option.label}
					</AppSelect.Option>
				))}
			</AppSelect>
		);
		const select = screen.getByTestId('app-select');
		expect(select).toBeInTheDocument();
		expect(wrapper).toMatchSnapshot();
	});

	it('should render AppSelect with options', () => {
		render(
			<AppSelect>
				{data.map(option => (
					<AppSelect.Option
						key={option.value}
						value={option.value}>
						{option.label}
					</AppSelect.Option>
				))}
			</AppSelect>
		);
		const options = screen.getAllByTestId('app-select-option');
		expect(options.length).toBe(4);
		options.forEach((option, index) => {
			expect(option).toHaveTextContent(data[index].label);
		});
	});

	it('should render with default option', () => {
		render(
			<AppSelect>
				{data.map(option => (
					<AppSelect.Option
						key={option.value}
						value={option.value}>
						{option.label}
					</AppSelect.Option>
				))}
			</AppSelect>
		);
		const options = screen.getAllByTestId('app-select-option') as HTMLOptionElement[];
		expect(options[0].selected).toBe(true);
	});

	it('should render AppSelect with options and select option', () => {
		render(
			<AppSelect>
				{data.map(option => (
					<AppSelect.Option
						key={option.value}
						value={option.value}>
						{option.label}
					</AppSelect.Option>
				))}
			</AppSelect>
		);
		const select = screen.getByTestId('app-select') as HTMLSelectElement;
		const options = screen.getAllByTestId('app-select-option') as HTMLOptionElement[];
		fireEvent.change(select, { target: { value: 2 } });
		expect(options[2].selected).toBe(true);
	});
});
