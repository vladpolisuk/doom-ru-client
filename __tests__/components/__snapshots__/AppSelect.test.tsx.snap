// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSelect should render AppSelect 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="app_select"
        data-testid="app-select"
      >
        <option
          class="app_select_option"
          data-testid="app-select-option"
          value=""
        >
          Choose option
        </option>
        <option
          class="app_select_option"
          data-testid="app-select-option"
          value="1"
        >
          Option 1
        </option>
        <option
          class="app_select_option"
          data-testid="app-select-option"
          value="2"
        >
          Option 2
        </option>
        <option
          class="app_select_option"
          data-testid="app-select-option"
          value="3"
        >
          Option 3
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="app_select"
      data-testid="app-select"
    >
      <option
        class="app_select_option"
        data-testid="app-select-option"
        value=""
      >
        Choose option
      </option>
      <option
        class="app_select_option"
        data-testid="app-select-option"
        value="1"
      >
        Option 1
      </option>
      <option
        class="app_select_option"
        data-testid="app-select-option"
        value="2"
      >
        Option 2
      </option>
      <option
        class="app_select_option"
        data-testid="app-select-option"
        value="3"
      >
        Option 3
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
